### DISCUSSION TITLE
Splitting FS from Module

### DISCUSSION THREAD

**Discussion by iliakan** [emscripten-core/emscripten#18721] [Ideas]
I recently tried to use PROXYFS to share files between modules and stumbled into two bugs that essentialy stop me from doing so. 

I consider refactoring Emscripten JS to move the FS object out of the module-creation function, and so allow such sharing.

Before I get down to this, how do you think, is there any show-stopper? Something that prevents FS to be shared across multiple modules (assuming -sMODULARIZE)? Any reason why each instance should have its own FS?

Maybe some issues with /dev or /proc?

It's crucial that FS doesn't reference Module, to allow one-to-many (one FS -  many modules) relation, is it so?

**Comment by sbc100** [comment]
I think fixing the bugs in PROXYFS is the probably the most logical way forwards.  Trying to actually share JS library code between modules is not something that emscripten was ever designed for so going down that path would likely be a lot of work.

**Comment by iliakan** [comment]
Answering my own question (because I saw similar ones): right now we have not only Module → FS relation (easy to fix by updating all properties of Module that reference FS and its methods)

There's also `mmap` method of FS, and it uses `mmapAlloc`, allocating memory of its enclosing `Module` and `HEAPs`. So there's FS → Module relation.

If I copy FS to another module, it's `mmap` still allocates memory in the old module (wrong).

I can also copy `mmapAlloc` to another module, but it calls other functions, which in turn use `HEAPU8` and other views, etc, connecting to the memory of the other module.
